function-controller/pkg/controller/autoscaler/autoscaler.go

```diff
@@ -0,0 +1,325 @@
+	requiredScaleDownProposals int
+	scalers                    map[FunctionId]scaler
+	replicas                   map[FunctionId]int // tracks all functions, including those which are not being monitored
+	maxReplicas                func(function FunctionId) int
+	delayScaleDown             func(function FunctionId) time.Duration
+	stop                       chan struct{}
+	accumulatingStopped        chan struct{}
+}
+
+// metrics counts the number of messages transmitted to a Subscription's topic and received by the Subscription.
+type metricsTotals struct {
+	transmitCount int32
+	receiveCount  int32
+}
+
+func (a *autoScaler) Propose() map[FunctionId]int {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+
+	proposals := make(map[FunctionId]int)
+	for _, funcTotals := range a.totals {
+		for fn, mt := range funcTotals {
+			proposals[fn] = a.scalers[fn](mt)
+			// Zero the sampled metrics for the next interval
+			funcTotals[fn] = &metricsTotals{}
+		}
+	}
+
+	return proposals
+}
+
+func (a *autoScaler) emptyQueue(funcId FunctionId) (bool, int64) {
+	for topic, funcTotals := range a.totals {
+		if _, ok := funcTotals[funcId]; ok {
+			queueLen, err := a.transportInspector.QueueLength(topic, funcId.Function)
+			if err != nil {
+				log.Printf("Failed to obtain queue length (and will assume it is positive): %v", err)
+				return false, -1
+			}
+			if queueLen > 0 {
+				return false, queueLen
+			}
+		}
+	}
+	return true, 0
+}
+
+func (a *autoScaler) StartMonitoring(topic string, fn FunctionId) error {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+	funcTotals, ok := a.totals[topic]
+	if !ok {
+		funcTotals = make(map[FunctionId]*metricsTotals)
+		a.totals[topic] = funcTotals
+	}
+
+	_, ok = funcTotals[fn]
+	if ok {
+		return fmt.Errorf("Already monitoring topic %s and function %s", topic, fn)
+	}
+
+	funcTotals[fn] = &metricsTotals{}
+
+	a.scalers[fn] = decorate(a.metricsScaler(fn), a.limitScalingUp(fn), a.limitScalingDown(fn), a.delay(fn))
+
+	return nil
+}
+
+func (a *autoScaler) delay(fn FunctionId) adjuster {
+	p := NewDelayer(func() time.Duration {
+		return a.delayScaleDown(fn);
+	})
+
+	return func(proposal int) int {
+		return p.Delay(proposal).Get()
+	}
+}
+
+func (a *autoScaler) StopMonitoring(topic string, function FunctionId) error {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+	funcTotals, ok := a.totals[topic]
+	if !ok {
+		return fmt.Errorf("Not monitoring topic %s and function %s", topic, function)
+	}
+
+	_, ok = funcTotals[function]
+	if !ok {
+		return fmt.Errorf("Not monitoring topic %s and function %s", topic, function)
+	}
+
+	delete(funcTotals, function)
+
+	// Avoid leaking memory.
+	if len(funcTotals) == 0 {
+		delete(a.totals, topic)
+	}
+	delete(a.scalers, function)
+
+	return nil
+}
+
+func (a *autoScaler) receiveLoop() {
+	producerMetrics := a.metricsReceiver.ProducerMetrics()
+	consumerMetrics := a.metricsReceiver.ConsumerMetrics()
+	for {
+		select {
+		case pm, ok := <-producerMetrics:
```

Comment:
who closes the metricsReceiver? Is `ok` ever gonna be false?

Reply:
I changed `receiveLoop` so it closes the metrics receiver on autoscaler close. Nevertheless, `ok` here should never be `false`, but this defensive coding seems reasonable. I've added a comment to make this clear.